wptrunner/executors/pytestrunner/fixtures.py

```diff
@@ -45,14 +46,79 @@ def setup(request, session):
         def test_something(setup, session):
             assert session.url == "https://example.org"
 
-    The session is closed when the test module goes out of scope by an
-    implicit call to `session.end`.
+    When the test function goes out of scope, any remaining user prompts
+    and opened windows are closed, and the current browsing context is
+    switched back to the top-level browsing context.
+
+    After returning the session to a near-pristine state, the session
+    is also closed by an implicit call to `session.end`.
     """
 
     def __init__(self, client):
         self.client = client
 
-    @pytest.fixture(scope="module")
+        # store handle of original window
+        # so that we do not close it
+        self.original_window = self.client.window_handle
+        self.client.end()
```

Comment:
I think that the finalizer added for client.end should take care of cleaning up everything at the end of each test, but you will end up not needing the restore_windows nor the switch_to_top_level_browsing_context if 'end' does it's job of closing all windows for the session (which should be a test in itself). Also I don't think you need dismiss_user_prompts as 'end' should handle that case too.... basically you don't need to do this change at all :)